soroban-env-host/src/host/comparison.rs

```diff
@@ -0,0 +1,274 @@
+            (Vec(a), Vec(b)) => self.compare(a, b),
+            (Map(a), Map(b)) => self.compare(a, b),
+            (Bytes(a), Bytes(b)) => self.as_budget().compare(&a.as_slice(), &b.as_slice()),
+            (ContractCode(a), ContractCode(b)) => self.as_budget().compare(a, b),
+            (AccountId(a), AccountId(b)) => self.as_budget().compare(a, b),
+
+            // List out at least one side of all the remaining cases here so
+            // we don't accidentally forget to update this when/if a new
+            // HostObject type is added.
+            (U64(_), _)
+            | (I64(_), _)
+            | (U128(_), _)
+            | (I128(_), _)
+            | (Vec(_), _)
+            | (Map(_), _)
+            | (Bytes(_), _)
+            | (ContractCode(_), _)
+            | (AccountId(_), _) => {
+                let a = host_obj_discriminant(a);
+                let b = host_obj_discriminant(b);
+                Ok(a.cmp(&b))
+            }
+        }
+    }
+}
+
+impl Compare<&[u8]> for Budget {
+    type Error = HostError;
+
+    fn compare(&self, a: &&[u8], b: &&[u8]) -> Result<Ordering, Self::Error> {
+        self.charge(CostType::BytesCmp, min(a.len(), b.len()) as u64)?;
+        Ok(a.cmp(b))
+    }
+}
+
+struct FixedSizeOrdType<'a, T: Ord>(&'a T);
+impl<T: Ord> Compare<FixedSizeOrdType<'_, T>> for Budget {
+    type Error = HostError;
+    fn compare(
+        &self,
+        a: &FixedSizeOrdType<'_, T>,
+        b: &FixedSizeOrdType<'_, T>,
+    ) -> Result<Ordering, Self::Error> {
+        self.charge(CostType::BytesCmp, core::mem::size_of::<T>() as u64)?;
+        Ok(a.0.cmp(&b.0))
+    }
+}
+
+macro_rules! impl_compare_fixed_size_ord_type {
+    ($t:ty) => {
+        impl Compare<$t> for Budget {
+            type Error = HostError;
+            fn compare(&self, a: &$t, b: &$t) -> Result<Ordering, Self::Error> {
+                self.compare(&FixedSizeOrdType(a), &FixedSizeOrdType(b))
+            }
+        }
+    };
+}
+
+impl_compare_fixed_size_ord_type!(u32);
+
+impl_compare_fixed_size_ord_type!(Hash);
+impl_compare_fixed_size_ord_type!(Uint256);
+impl_compare_fixed_size_ord_type!(ScContractCode);
+impl_compare_fixed_size_ord_type!(AccountId);
+impl_compare_fixed_size_ord_type!(PublicKey);
+impl_compare_fixed_size_ord_type!(TrustLineAsset);
+
+impl_compare_fixed_size_ord_type!(LedgerKeyAccount);
+impl_compare_fixed_size_ord_type!(LedgerKeyTrustLine);
+impl_compare_fixed_size_ord_type!(LedgerKeyOffer);
+impl_compare_fixed_size_ord_type!(LedgerKeyData);
+impl_compare_fixed_size_ord_type!(LedgerKeyClaimableBalance);
+impl_compare_fixed_size_ord_type!(LedgerKeyLiquidityPool);
+impl_compare_fixed_size_ord_type!(LedgerKeyContractCode);
+impl_compare_fixed_size_ord_type!(LedgerKeyConfigSetting);
+
+impl_compare_fixed_size_ord_type!(LedgerEntryExt);
+
+impl_compare_fixed_size_ord_type!(AccountEntry);
+impl_compare_fixed_size_ord_type!(TrustLineEntry);
+impl_compare_fixed_size_ord_type!(OfferEntry);
```

Comment:
Do we need all of these? For example, when will we compare (or even use) offer entries in Soroban?

Reply:
Well, they're substructures of LedgerKey / LedgerEntry which are stored in the storage maps. I guess since we're just charging at the base level of them and not descending into their substructure I can eliminate the substructure cases.

Reply:
(Sorry, wrote a wrong comment here): they're substructures of LedgerKey / LedgerEntry and we explicitly switch on the outer structure to special-case the charge for the one variable-sized case: `ContractData`. So we need to also handle the fixed-size cases, even if they never get stored in practice.)